### PR TITLE
Buffered query result POC

### PR DESCRIPTION
**Pull request by Tishj**
Save finished datachunks and limit how many buffers there are by blocking the Sink of the result collector

### CODE REVIEW

File: src/common/vector_operations/vector_copy.cpp [hunk 1]
```diff
@@ -34,11 +34,10 @@ static const ValidityMask &CopyValidityMask(const Vector &v) {
 	}
 }
 
-void VectorOperations::Copy(const Vector &source_p, Vector &target, const SelectionVector &sel_p, idx_t source_count,
-                            idx_t source_offset, idx_t target_offset) {
+void VectorOperations::CopyPartial(const Vector &source_p, Vector &target, const SelectionVector &sel_p,
+                                   idx_t source_count, idx_t source_offset, idx_t copy_count, idx_t target_offset) {
```
Comment (Tishj) [inline src/common/vector_operations/vector_copy.cpp:38]: Had to copy an X amount of tuples from `source`, not `source_count - source_offset`

File: src/include/duckdb/execution/operator/helper/physical_batch_collector.hpp [hunk 2]
```diff
@@ -33,10 +33,6 @@ class PhysicalBatchCollector : public PhysicalResultCollector {
 		return true;
 	}
 
-	bool IsSink() const override {
-		return true;
```
Comment (samansmink) [inline src/include/duckdb/execution/operator/helper/physical_batch_collector.hpp:37]: is this on puropose?

File: src/parallel/executor.cpp [hunk 3]
```diff
@@ -442,6 +442,32 @@ void Executor::RescheduleTask(shared_ptr<Task> &task_p) {
 	}
 }
 
+bool Executor::ResultCollectorIsBlocked() const {
+	if (completed_pipelines + 1 != total_pipelines) {
+		// The result collector is always in the last pipeline
+		return false;
+	}
+	if (to_be_rescheduled_tasks.empty()) {
```
Comment (samansmink) [inline src/parallel/executor.cpp:450]: Not sure, but could this be racy in some situations? Like when a task hasn't been added to `to_be_rescheduled_tasks`? I guess for now we always hold the executor lock when calling this so its fine?

**Review by samansmink — COMMENTED** [review]
Just scanned through it briefly, I think it looks good in general!

File: src/include/duckdb/execution/operator/helper/physical_batch_collector.hpp (hunk 2 above)
Comment (Tishj) [reply src/include/duckdb/execution/operator/helper/physical_batch_collector.hpp:37]: Yea `PhysicalResultCollector` already overrides IsSink to true, so this is redundant as far as I can tell.

File: src/parallel/executor.cpp (hunk 3 above)
Comment (Tishj) [reply src/parallel/executor.cpp:450]: Hmm fair, it's a private method though so I would argue it's expected the lock is held when this is called

**Pull request merged**